### ISSUE TITLE
Konzeptaufgabe: Factory-Pattern

### ISSUE THREAD

**Issue by AMatutat** [PM-Dungeon/PM-Lecture#104]

**Comment by AHeinisch** [comment]
Da hier was neues entwickelt werden muss in welchen umfang soll dies sein also eigentlich ist dieses Pattern ja dafür da Objekte zu erstellen. 
Also könnte eine Aufgabe sein Hier habt ihr 1 Klasse welche keinen Konstruktor hat um alle Felder zu initialisieren implementieren hierfür eine Factory Methode welche alle Felder initialisiert.
Frage dazu wäre soll ansonsten bei dem Pattern noch was abgefragt werden ?

**Comment by cagix** [comment]
genau, das pattern bietet eine zentrale methode, die die eigentlichen objekte erstellt. dabei können das einerseits objekte des selben typs sein, nur mit unterschiedlich befüllten attributen oder andererseits auch verschiedene objekte/klassen, die von einem gemeinsamen obertyp ableiten. -- ich würde beides gern sehen.

in der vorlesung mache ich immer das beispiel mit dem ticketautomat, wo ich als user nur mein fahrtziel (und vielleicht noch weitere eigenschaften wie 1. oder 2. klasse) eingebe und der automat daraufhin das passende ticket (also objekt der passenden unterklasse) erstellt.

sowas in der art, nur halt nicht ticketautomat :)

**Comment by cagix** [comment]
szenario: kunde kommt bestellt einen rechner und gibt an, was er damit machen möchte (nutzung am schreibtisch vs. nutzung unterwegs, office und mails vs. entwicklung, ...). die methode soll daraufhin die passende sub-klasse instantiieren (laptop, desktop, cloud) und darin die attribute korrekt setzen (cpu/kerne, ram, ssd, ...).

**Comment by cagix** [comment]
szenaria: kunde bestellt ein fahrrad und gibt an, was er damit machen möchte (pendeln zur arbeit, mitnahme im zug, schnell auf der straße fahren, auch mal nen waldweg fahren, richtig durch den acker pflügen). und er gibt an, wie groß und schwer er ist und welche farbe er am liebsten mag. das system erstellt ein objekt der passenden unterklasse (obertyp fahrrad, subklassen brompton, stadtgurke, rennrad, gravel, mtb). und mit den zusätzlichen angaben zur größe und farbe wird dann die rahmenhöhe und vorbaulänge und reifenbreite (innerhalb der typischen grenzen) eingestellt und die rahmenfarbe ausgewählt.

**Comment by cagix** [comment]
@AHeinisch zwei bitten: bitte assigne das ticket an dich, wenn du das hier bearbeiten möchtest. und sprich gezielt @AMatutat oder mich in den fragen an, so lange ich nicht in cc im issue bin. habe deine frage nur zufällig gesehen ...

**Comment by AHeinisch** [comment]
@cagix ich glaube das mit den assignen hatten wir noch nicht gefixed daher klappt das leider noch nicht vielleicht kann das @AMatutat einmal schauen wo das dran liegen könnte. Mir fehlen hierfür wahrscheinlich einfach die rechte um mir das Issue zu nehmen.
![image](https://user-images.githubusercontent.com/32952411/159433418-4cd6e95e-7212-4895-977f-922b9f244271.png)

**Comment by cagix** [comment]
> @cagix ich glaube das mit den assignen hatten wir noch nicht gefixed daher klappt das leider noch nicht vielleicht kann das @AMatutat einmal schauen wo das dran liegen könnte. Mir fehlen hierfür wahrscheinlich einfach die rechte um mir das Issue zu nehmen.
> ![image](https://user-images.githubusercontent.com/32952411/159433418-4cd6e95e-7212-4895-977f-922b9f244271.png)

oh f*ck. ok, da muss ich mal schauen. das ist wieder so eine berechtigungskiste :/

**Comment by cagix** [comment]
@AHeinisch schau mal bitte nochmal, sollte jetzt gehen ...

dummerweise geht das nur im komplettpack einher mit du darfst jetzt auch issues löschen und so ... also nutze deine neuen kräfte weise :)

**Comment by AHeinisch** [comment]
Ja jetzt habe ich die rechte dazu mich zuzuweisen. Somit habe ich das doch mal gemacht. danke @cagix

**Comment by cagix** [comment]
@AHeinisch Was ist denn hier der Stand? Wir brauchen die Aufgabe nächste Woche.

**Comment by jedi101** [comment]
@AHeinisch wenn Du magst kann ich Dir dafür vielleicht auch zuarbeiten/etwas abnehmen? Sonst bekomme ich wieder Strafpunkte von Carsten oder so.

**Comment by cagix** [comment]
> @AHeinisch wenn Du magst kann ich Dir dafür vielleicht auch zuarbeiten/etwas abnehmen? Sonst bekomme ich wieder Strafpunkte von Carsten oder so.

@jedi101 da verwechselst du was. bei mir gibt es keine strafpunkte -- du wirst direkt zum ruderdienst eingeteilt ;)  /s

**Comment by cagix** [comment]
@jedi101 vielleicht wäre es doch gut, wenn du dich da nochmal dran setzt. wir brauchen das morgen fertig, und @AHeinisch reagiert nicht.

**Comment by jedi101** [comment]
@cagix - Siehe: https://github.com/PM-Dungeon/Alex/pull/6

**Comment by AHeinisch** [comment]
Es tut mir leid, dass ich aktuell so neben der Spur bin. Ist dieses Issue überhaupt fertig  @jedi101 ? Ich glaube, der Workflow ist etwas anstrengend auf Dauer mit erst in einem Repo arbeiten welches eigentlich identisch zu Solutions ist. Dann auf die drei Repos verteilen. 

Aber mal eine Frage soll das Issue geschlossen werden bevor der gesamte Workflow abgearbeitet ist oder erst sobald alle Schritte durchgeführt wurden @cagix .

**Comment by cagix** [comment]
> Es tut mir leid, dass ich aktuell so neben der Spur bin. 

Kein Thema. Nur bitte immer etwas kommunizieren, wenn das geht.

> Ist dieses Issue überhaupt fertig @jedi101 ? Ich glaube, der Workflow ist etwas anstrengend auf Dauer mit erst in einem Repo arbeiten welches eigentlich identisch zu Solutions ist. Dann auf die drei Repos verteilen.

Ich denke, das hat @jedi101 "drüben" in Deinem Repo fertig gemacht und kippt das jetzt in die drei Zielrepos ein. Insofern sollte hier fertig sein.

Und ja, der Workflow ist nicht wirklich gut. Das ist zum Teil gewachsen und zum Teil stammt das aus ner Zeit, als wir noch andere Strukturen vor Augen hatten. Ich habe mir schon #345 und #353 angelegt, um nochmal gründlich über die Strukturen nachzudenken und dabei auch die Workflows mitzudenken. @AHeinisch Vorschläge sind in den Tickets gern willkommen.


> 
> Aber mal eine Frage soll das Issue geschlossen werden bevor der gesamte Workflow abgearbeitet ist oder erst sobald alle Schritte durchgeführt wurden? @cagix

@AHeinisch Der Aufgabentext wird ja noch per PR hier ins Lecture-Repo eingekippt. Dieser PR sollte mit dem Ticket verknüpft werden, damit es zugeht, sobald die Aufgabe im Repo ist. Ich glaube, das war bei einem anderen Ticket (wo Du drin stehst) der Grund für meine Nachfrage, was wohl der Stand ist ...
